### ISSUE TITLE
auto discover should read `name` attribute from Django AppConfigs

### ISSUE THREAD

**Issue by stringfellow** [celery/celery#2596]
The AppConfig module is not always in the same package as the named app (https://docs.djangoproject.com/en/1.8/ref/applications/#for-application-users), but the `name` attribute is always present in that config object (https://docs.djangoproject.com/en/1.8/ref/applications/#django.apps.AppConfig.name).

The `find_related_module` method should check that attribute.

References #2248

**Comment by ask** [comment]
autodiscover_tasks does not know anything about INSTALLED_APPS, so the user would have to
use:

    from django.apps import apps
    app.autodiscover_tasks(lambda: [n.name for n in apps.get_app_configs()]

There's a much nicer solution, but sadly `apps.get_apps` is deprecated and will be removed in Django 1.9 for some mysterious reason:

    from django.apps import apps
    app.autodiscover_tasks(apps.get_apps)

**Comment by stringfellow** [comment]
Aha! Awesome :) Thanks!

**Comment by danxshap** [comment]
@ask am I correct in understanding that the following is the current best practice for 3.1.18?

```
from django.apps import apps
app.autodiscover_tasks(lambda: [n.name for n in apps.get_app_configs()])
```

Per the docs, I originally had it pulling from INSTALLED_APPS:
```
app.autodiscover_tasks(lambda: settings.INSTALLED_APPS)
```

But then I started using custom AppConfig subclasses for some of my apps (with the paths to those AppConfig subclasses in INSTALLED_APPS), and once I did that, those apps' tasks no longer became registered. I needed to switch to the first example above to get it to work.

**Comment by ask** [comment]
Yeah, I think so as soon as you use custom AppConfigs in a way that this matters, then you have to get them from there.

It would be cool if autodiscover_tasks could also support a list of modules as argument, that way we
could use:

    app.autodiscover_tasks(apps.get_apps)

instead

**Comment by pannal** [comment]
This isn't merged in any current release yet, is it?

**Comment by ask** [comment]
It's being included in 4.0, you can only use the workaround above for now.
